sys-fs/lvm2/files/lvm.rc-r2

```diff
@@ -0,0 +1,174 @@
+#!/sbin/openrc-run
+# Copyright 1999-2020 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+_get_lvm_path() {
+	local lvm_path=
+	for lvm_path in /bin/lvm /sbin/lvm ; do
+		[ -x "${lvm_path}" ] && break
+	done
+	echo "${lvm_path}"
+}
+
+_use_lvmetad() {
+	local lvm_path="$(_get_lvm_path)"
+	[ ! -x "${lvm_path}" ] && return 1
+	${lvm_path} dumpconfig global 2>/dev/null | grep -q 'use_lvmetad=1'
+}
+
+_use_lvmlockd() {
+	local lvm_path="$(_get_lvm_path)"
+	[ ! -x "${lvm_path}" ] && return 1
+	${lvm_path} dumpconfig global 2>/dev/null | grep -q 'use_lvmlockd=1'
+}
+
+depend() {
+	before checkfs fsck
+	after modules device-mapper
+	# We may want lvmetad based on the configuration. If we added lvmetad
+	# support while lvm2 is running then we aren't dependent on it. For the
+	# more common case, if its disabled in the config we aren't dependent
+	# on it.
+	config /etc/lvm/lvm.conf
+	local _want=
+
+	if service_started ; then
+		_want=$(service_get_value want)
+	else
+		if _use_lvmetad ; then
+			_want="${_want} lvmetad"
+		fi
+
+		if _use_lvmlockd ; then
+			_want="${_want} lvmlockd"
+		fi
+	fi
+
+	# Make sure you review /etc/conf.d/lvm as well!
+	# Depending on your system, it might also introduce udev & mdraid
+	need sysfs
+
+	if [ -n "${_want}" ] ; then
+		want ${_want}
+	fi
+}
+
+config='global { locking_dir = "/run/lock/lvm" }'
+
+dm_in_proc() {
+	local retval=0
+	for x in devices misc ; do
+		grep -qs 'device-mapper' /proc/${x}
+		retval=$((${retval} + $?))
+	done
+	return ${retval}
+}
+
+start() {
+	# LVM support for /usr, /home, /opt ....
+	# This should be done *before* checking local
+	# volumes, or they never get checked.
+
+	# NOTE: Add needed modules for LVM or RAID, etc
+	#       to /etc/modules.autoload if needed
+
+	lvm_path="$(_get_lvm_path)"
+	if [ -z "${lvm_path}" ] ; then
+		eerror "Failed to find lvm binary in /bin or /sbin!"
+		return 1
+	fi
+
+	if [ -z "${CDBOOT}" ] ; then
+		if [ -e /proc/modules ] && ! dm_in_proc ; then
+			ebegin "Trying to load dm-mod module"
+			modprobe dm-mod 2>/dev/null
+			eend $?
+		fi
+
+		if [ -d /proc/lvm ] || dm_in_proc ; then
+			local has_errors=0 verbose_command mknodes=""
+
+			yesno "${rc_verbose}" && verbose_command=" -v"
+
+			ebegin "Starting the Logical Volume Manager"
+
+			if _use_lvmetad ; then
+				# Extra PV find pass because some devices might not have been available until very recently
+				${lvm_path} pvscan${verbose_command} --config "${config}" --cache
+				[ $? -ne 0 ] && has_errors=1
+			fi
+
+			# Now make the nodes
+			[ -S /run/udev/control ] || mknodes=" --mknodes"
```

Comment:
The socket might exist even if udevd has exited. A more robust test would be something like:
```suggestion
			udevadm control --ping --timeout=5 || mknodes=" --mknodes"
```

Reply:
Assuming udevadm command is available ... which it may not be in some truly obscure cases.  There has also been a suggestion that the root cause may be fixed upstream already anyway, so busy investigating that, otherwise I'll see what I can do about using udevadm control, perhaps chain the whole lot so only even try udevadm control --ping *if* the socket exists kind of thing so we can have a higher degree of certainty that udevadm command will be available.

Reply:
Create a thin pool + volume:

```
arthur [12:05:17] ~ # lvcreate -T -L100G --name thin lvm
  Thin pool volume with chunk size 64.00 KiB can address at most <15.88 TiB of data.
  Logical volume "thin" created.
arthur [12:05:22] ~ # lvcreate --virtualsize 50G --thinpool lvm/thin --name testing
  Logical volume "testing" created.
```

At this point we expect vgscan with --mknodes to fail on older lvm2 version:

```
arthur [12:06:09] ~ # equery list -ipo lvm2
 * Searching for lvm2 ...
[IP-] [  ] sys-fs/lvm2-2.03.21-r1:0
[-P-] [ ~] sys-fs/lvm2-2.03.22-r1:0
[-P-] [  ] sys-fs/lvm2-2.03.22-r3:0
arthur [12:06:13] ~ # /sbin/lvm vgscan --config 'global { locking_dir = "/run/lock/lvm" }' --mknodes  Found volume group "lvm" using metadata type lvm2
  Command failed with status code 5.
arthur [12:06:32] ~ # echo $?
5
```

So we have that at least.  Reproduced.  Upgrade to newest lvm2 and we have:

```
arthur [12:10:07] ~ # equery list -ipo lvm2
 * Searching for lvm2 ...
[-P-] [  ] sys-fs/lvm2-2.03.21-r1:0
[-P-] [ ~] sys-fs/lvm2-2.03.22-r1:0
[IP-] [  ] sys-fs/lvm2-2.03.22-r3:0
arthur [12:10:15] ~ # /sbin/lvm vgscan --config 'global { locking_dir = "/run/lock/lvm" }' --mknodes  Found volume group "lvm" using metadata type lvm2
arthur [12:10:17] ~ # echo $?
0
```

Based on this, I believe we can close both relevant bugs as "fixed upstream" and I'll close this PR.

Reply:
I think you misunderstood what I was trying to say.  I was saying that it's possible to encounter a system without udevd (theoretically, although extremely unlikely).  At that point it becomes very hard to run udevadm ... at which point first check for the socket, if the socket doesn't exist, use --mknodes, if the socket exist (udev is probably running, and the commands is available) and udevadm fails use --mknodes, if the socket exist and udevadm succeeds, don't use --mknodes.

Hope that more clearly explains what my intention was.

But all of that is now pointless.

Reply:
If udev is not installed, calling udevadm will fail and we get the same result.

Checking for the control socket is entirely redundant.